library/common.py

```diff
@@ -5,25 +5,120 @@
 from urllib.parse import urljoin
 from urllib.parse import quote
 
-class ProvisionerError(Exception):
-    def __init__(self, message):
-        super(ProvisionerError, self).__init__(message)
-
 # TODO: This needs a more standard name (static method)
 # Or we need to make sure the IPGetter object is easy to use
-def get_machine_by_name(mrp_token, mrp_url, machine_name):
+def get_machine_by_name(requester, machine_name):
     """ Look up machine by name """
-    headers = {'Authorization': mrp_token}
     q = '(= name "{}")'.format(quote(machine_name))
-    url = urljoin(mrp_url, "/api/v1/machine?q={}&show_all=false".format(q))
-    r = requests.get(url, headers=headers)
-    if r.status_code != 200:
-        raise ProvisionerError('Error fetching {}, HTTP {} {}'.format(mrp_url,
-                         r.status_code, r.reason))
-    if len(r.json()) == 0:
-       raise ProvisionerError('Error no assigned machine found with name "{}"'.
-                    format(machine_name))
-    if len(r.json()) > 1:
-       raise ProvisionerError('Error more than one machine found with name "{}", {}'.
-                    format(machine_name, r.json()))
-    return r.json()[0]
+    path = "/api/v1/machine?q={}&show_all=false".format(q)
+    try:
+        result = requester.get(path)
+    except Exception as err:
+        raise err
+
+    return result[0]
+
+class ProvisionerError(Exception):
+    def __init__(self, message):
+        super(ProvisionerError, self).__init__(message)
+
+class ClientError(Exception):
+    def __init__(self, message):
+        super(ClientError, self).__init__(message)
+
+class RequesterError(Exception):
+    def __init__(self, message, method, url):
+        super(RequesterError, self).__init__(message)
+        self.method = method
+        self.url = url
+
+class RequesterHTTPError(RequesterError):
+    def __init__(self, method, url, status_code, response):
+        super(RequesterHTTPError, self).__init__("HTTP ERROR %s on %s request at %s" %
+                                                 (str(status_code), str(method), str(url)),
+                                                 method, url)
+        self.status_code = status_code
+        self.response = response
+
+class RequesterJSONError(RequesterError):
+    def __init__(self, method, url, response):
+        super(RequesterJSONError, self).__init__("JSON ERROR on %s request at %s" %
+                                                 (str(method), str(url)), method, url)
+        self.response = response
+
+
+class Requester(object):
+    def __init__(self, mrp_url, mrp_token):
+        self.base_url = mrp_url
```

Comment:
validation on these fields would make for better error reports.

Reply:
Now the Requester 'pings' MrP at initialization.